pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/AvgShedder.java

```diff
@@ -0,0 +1,320 @@
+                brokerHitCountForHigh.remove(maxBroker);
+                brokerHitCountForHigh.remove(minBroker);
+
+                brokerHitCountForLow.remove(maxBroker);
+                brokerHitCountForLow.remove(minBroker);
+            } else {
+                pairs.add(Pair.of(minBroker, maxBroker));
+                if (brokerScoreMap.get(maxBroker) - brokerScoreMap.get(minBroker) < highThreshold) {
+                    brokerHitCountForLow.computeIfAbsent(minBroker, k -> new MutableInt(0)).increment();
+                    brokerHitCountForLow.computeIfAbsent(maxBroker, k -> new MutableInt(0)).increment();
+
+                    brokerHitCountForHigh.remove(maxBroker);
+                    brokerHitCountForHigh.remove(minBroker);
+                } else {
+                    brokerHitCountForLow.computeIfAbsent(minBroker, k -> new MutableInt(0)).increment();
+                    brokerHitCountForLow.computeIfAbsent(maxBroker, k -> new MutableInt(0)).increment();
+
+                    brokerHitCountForHigh.computeIfAbsent(minBroker, k -> new MutableInt(0)).increment();
+                    brokerHitCountForHigh.computeIfAbsent(maxBroker, k -> new MutableInt(0)).increment();
+                }
+            }
+            i++;
+            j--;
+        }
+        return pairs;
+    }
+
+    @Override
+    public Optional<String> selectBroker(Set<String> candidates, BundleData bundleToAssign, LoadData loadData,
+                                         ServiceConfiguration conf) {
+        if (!bundleBrokerMap.containsKey(bundleToAssign) || !candidates.contains(bundleBrokerMap.get(bundleToAssign))) {
+            // cluster initializing or broker is shutdown
+            if (log.isDebugEnabled()) {
+                if (!bundleBrokerMap.containsKey(bundleToAssign)) {
+                    log.debug("cluster is initializing");
+                } else {
+                    log.debug("expected broker:{} is shutdown, candidates:{}", bundleBrokerMap.get(bundleToAssign),
+                            candidates);
+                }
+            }
+            String broker = getExpectedBroker(candidates, bundleToAssign);
+            bundleBrokerMap.put(bundleToAssign, broker);
+            return Optional.of(broker);
+        } else {
+            return Optional.of(bundleBrokerMap.get(bundleToAssign));
+        }
+    }
+
+    private static String getExpectedBroker(Collection<String> brokers, BundleData bundle) {
+        List<String> sortedBrokers = new ArrayList<>(brokers);
+        Collections.sort(sortedBrokers);
+
+        try {
+            // use random number as input of hashing function to avoid special case that,
+            // if there is 4 brokers running in the cluster,and add broker5,and shutdown broker3,
+            // then all bundles belonging to broker3 will be loaded on the same broker.
+            final long hashcode = Hashing.crc32().hashString(String.valueOf(random.nextInt()),
```

Comment:
It would be better to use `new Random()` here instead of a static `Random` object whose random seed is fixed. Though it won't make a big difference.

BTW, `getExpectedBroker` is never covered by any test, could you add a unit test for this case?

Reply:
Updated.